woocommerce/woocommerce-ios
# Product List: Extract the product category list view and view model to their own classes.

Part of https://github.com/woocommerce/woocommerce-ios/issues/5159
Sorry for the more than 500 lines of code changing, most of the changes are moving the logic to new classes.

## Description
In the context of https://github.com/woocommerce/woocommerce-ios/issues/5159, we want to extract the product category list view used for editing or adding a product into its own dedicated class so it can be reused when filtering products by category.
That way we will avoid code duplication for the product categories presentation and fetching logic.

In this. manner, we create a new `EditProductCategoryListViewController` and `EditProductCategoryListViewModel` to replicate the logic previously implemented by `ProductCategoryListViewController` and `ProductCategoryListViewModel`, composing the product list with the add button and the specific use case logic.

## Changes
- Extract product category list view to `ProductCategoryListViewController`
- Extract product category list presentation logic to `ProductCategoryListViewModel` Since it is not related to a specific product, we just need the `siteID` to operate
- Use view controller containment in `EditProductCategoryListViewController` to show the product category list together with the specific case UI (Add Category Button)
- Create `EditProductCategoryListViewModel`, that handles the business logic of this case holding a reference to the product category list view model
- Split unit tests files accordingly
- Reorganize the project group structure to accommodate this new implementation

<img src="https://user-images.githubusercontent.com/1864060/137512964-212b7223-af64-42d2-bacd-82f6e3c9ce5e.png" width="320" height="693">

## Testing
Since no new functionality is added here, we have to make sure that the product categories functionality works as before
Prerequisites: your account has to have one store, ideally with products and categories.

Use Case 1: Assign  a existing category to a product

- Open the app, go to the Products tab
- Select a product
- Select Categories in the menu
- Select any category and press Done, it should be added to the Product.

Use Case 2: Unassign a category from a product

- Open the app, go to the Products tab
- Select a product
- Make sure that it has categories assigned
- Select Categories in the menu
- Make sure that these categories are checked when opening the categories list
- Uncheck that category
- Make sure that that category is not assigned to the product anymore

Use Case 3: Create new category and add it to the product

- Open the app, go to the Products tab
- Select a product
- Select Categories in the menu
- Press add category
- Create a new category following the steps in the menu
- Make sure that that category is checked in the product category list
- When pressing Done, make sure that that category is assigned to the product

Use Case 3: Create new product and assign a category

- Open the app, go to the Products tab
- Press on the + button, at the top right
- Select a Product type
- Press on "+ Add more details"
- Select Categories
- Existing categories should be shown
- Check any category
- When pressing Done, make sure that that category is assigned to the product




Update release notes:
